org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/SnippetCompletionProposal.java

```diff
@@ -46,32 +46,97 @@
 	private static final String CLASS_SNIPPET_LABEL = "class";
 	private static final String INTERFACE_SNIPPET_LABEL = "interface";
 	private static final String CLASS_KEYWORD = "class";
 	private static final String INTERFACE_KEYWORD = "interface";
-	private static final Set<String> UNSUPPORTED_RESOURCES = Sets.newHashSet("module-info.java", "package-info.java");
 
 	private static String PACKAGEHEADER = "package_header";
 	private static String CURSOR = "cursor";
 
-	public static List<CompletionItem> getSnippets(ICompilationUnit cu, CompletionContext completionContext, IProgressMonitor monitor) {
+	public static List<CompletionItem> getSnippets(ICompilationUnit cu, CompletionContext completionContext, IProgressMonitor monitor) throws JavaModelException {
 		if (cu == null) {
 			throw new IllegalArgumentException("Compilation unit must not be null"); //$NON-NLS-1$
 		}
+
+		List<CompletionItem> res = new ArrayList<>();
+		res.addAll(getGenericSnippets(cu, completionContext));
+		res.addAll(getClassAndInterfaceSnippets(cu, completionContext, monitor));
+
+		return res;
+	}
+
+	private static List<CompletionItem> getGenericSnippets(ICompilationUnit cu, CompletionContext completionContext) throws JavaModelException {
+		List<CompletionItem> res = new ArrayList<>();
+
+		char[] completionToken = completionContext.getToken();
+		if (completionToken == null) {
+			return Collections.emptyList();
+		}
+		int tokenLocation = completionContext.getTokenLocation();
+		JavaContextType contextType = (JavaContextType) JavaLanguageServerPlugin.getInstance().getTemplateContextRegistry().getContextType(JavaContextType.ID_STATEMENTS);
+		if (contextType == null) {
+			return Collections.emptyList();
+		}
+
+		IDocument document = new Document(cu.getSource());
+		DocumentTemplateContext javaContext = contextType.createContext(document, completionContext.getOffset(), completionToken.length, cu);
+		Template[] templates = null;
+		if ((tokenLocation & CompletionContext.TL_STATEMENT_START) != 0) {
+			templates = JavaLanguageServerPlugin.getInstance().getTemplateStore().getTemplates(JavaContextType.ID_STATEMENTS);
+		} else {
+			// We only support statement templates for now.
+		}
+
+		if (templates == null || templates.length == 0) {
+			return Collections.emptyList();
+		}
+
+		for (Template template : templates) {
+			if (!javaContext.canEvaluate(template)) {
+				continue;
+			}
+			TemplateBuffer buffer = null;
+			try {
+				buffer = javaContext.evaluate(template);
+			} catch (BadLocationException | TemplateException e) {
+				JavaLanguageServerPlugin.logException(e.getMessage(), e);
+				continue;
+			}
+			if (buffer == null) {
+				continue;
+			}
+			String content = buffer.getString();
+			if (Strings.containsOnlyWhitespaces(content)) {
+				continue;
+			}
+			final CompletionItem item = new CompletionItem();
+			item.setLabel(template.getName());
+			item.setInsertText(content);
+			item.setDetail(template.getDescription());
```

Comment:
that's where you need to sanitize descriptions, remove `${\d}` and replace `${\d:(.*)}`  with the matching group(1)

Reply:
(regexp syntax not verified/guaranteed)

Reply:
Updated the snippet document with the following change:
- Escape the cursor placeholder
- Render the document as `MarkDown` format if the client side supports

Some screenshots:
![WeChat Screenshot_20191012105314](https://user-images.githubusercontent.com/6193897/66694144-1f65b000-ece3-11e9-82ff-9721ba0a9ede.png)
![WeChat Screenshot_20191012105327](https://user-images.githubusercontent.com/6193897/66694145-1f65b000-ece3-11e9-8671-41a8f246f5b5.png)
![WeChat Screenshot_20191012105336](https://user-images.githubusercontent.com/6193897/66694146-1ffe4680-ece3-11e9-8d7b-13fc185de0a2.png)

Meanwhile, the document of the class and interface snippets also benefit from this change:

#### Before:
![image](https://user-images.githubusercontent.com/6193897/66694166-47edaa00-ece3-11e9-8912-e5beefd0174a.png)


#### After:
![WeChat Screenshot_20191012105353](https://user-images.githubusercontent.com/6193897/66694147-1ffe4680-ece3-11e9-9139-d060bd549d77.png)